### PR TITLE
feat: Add databuilder models for Feature

### PR DESCRIPTION
**Pull request by dkunitsk**
Signed-off-by: Dmitriy Kunitskiy <dkunitskiy@lyft.com>

### Summary of Changes
This PR adds `FeatureMetadata`, `Feature_Generation_Code`, and `Feature_Watermark` models. Like dashboards, the feature models are "normalized" so that core metadata is ingested first via the `FeatureMetadata` model and additional metadata is added on top. 

A new generic `Owner` model is added which takes a `start_label`, `start_key`. Both table owner and dashboard owner are re-written to inherit from the new generic `Owner` model, resulting in boilerplate code cleanup.

The `Tag`, `Badge`, `Owner`, and `Programmatic_Description` work with features and in some cases have needed minor changes to be more generic. 

Care has been taken to keep this PR backwards compatible.

### Tests

- Added tests for all new models. Modified tests of other models which were made more general to work with features. 

### Documentation

- `models.md` describes new Feature models

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

File: databuilder/databuilder/models/feature/feature_query.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+from databuilder.models.feature.feature_metadata import FeatureMetadata
+from databuilder.models.graph_node import GraphNode
+from databuilder.models.graph_relationship import GraphRelationship
+from databuilder.models.graph_serializable import GraphSerializable
+
+
+class FeatureQuery(GraphSerializable):
+    NODE_LABEL = 'Feature_Query'
```
Comment (dkunitsk) [inline databuilder/databuilder/models/feature/feature_query.py:15]: Rename to Feature_Generation_Code.

File: databuilder/databuilder/models/feature/feature_metadata.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+class FeatureMetadata(GraphSerializable):
+    """
+    Base feature metadata.
+
+    It implements GraphSerializable (TODO: implement TableSerializable)
+    so that it can be serialized to produce Feature, Feature_Group, Tag,
+    Database, Description and relations between those.
+
+    Other models (created independently) which may be related to Feature are:
+      - FeatureQuery
```
Comment (allisonsuarez) [inline databuilder/databuilder/models/feature/feature_metadata.py:26]: nit: feature query changed name

File: databuilder/databuilder/models/feature/feature_watermark.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+from typing import Iterator, Union
+
+from databuilder.models.feature.feature_metadata import FeatureMetadata
+from databuilder.models.graph_node import GraphNode
+from databuilder.models.graph_relationship import GraphRelationship
+from databuilder.models.graph_serializable import GraphSerializable
+from databuilder.models.timestamp.timestamp_constants import TIMESTAMP_PROPERTY
+
+
+class FeatureWatermark(GraphSerializable):
```
Comment (allisonsuarez) [inline databuilder/databuilder/models/feature/feature_watermark.py:13]: Maybe add comment on why we distinguish from table watermark?

File: databuilder/tests/unit/models/feature/test_feature_generation_code.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Copyright Contributors to the Amundsen project.
+# SPDX-License-Identifier: Apache-2.0
+
+import unittest
+
+from databuilder.models.feature.feature_generation_code import FeatureGenerationCode
+from databuilder.serializers import neo4_serializer
+
+
+class TestFeatureGenerationCode(unittest.TestCase):
```
Comment (allisonsuarez) [inline databuilder/tests/unit/models/feature/test_feature_generation_code.py:10]: Same little comment explanation of why this is not just Query like for dashboards

**Review by allisonsuarez — APPROVED** [review]
LGTM! Left a couple nits but otherwise this looks great

**Comment by dkunitsk** [comment]
@dorianj @verdan are one of you interested in having a look?

**Pull request merged**